zerodb/catalog/indexes/field.py

```diff
@@ -30,10 +76,166 @@ def scan_forward(self, docids, limit=None):
+                for docid in curdocids:
+                    if docid in docids:
+                        n += 1
+                        yield docid
+                        if limit and n >= limit:
+                            raise StopIteration
+
+    def index_doc(self, docid, obj):
+        if self.discriminator_callable:
+            # Model class definition has a list of virtual fields
+            virtuals = getattr(obj.__class__, "_z_virtual_fields", {})
+            value = virtuals.get(self.discriminator, _marker)
+            if value != _marker:
+                try:
+                    value = value(obj)
+                except:
+                    value = _marker
+        else:
+            value = getattr(obj, self.discriminator, _marker)
+
+        if value is _marker:
+            # unindex the previous value
+            super(CatalogIndex, self).unindex_doc(docid)
+
+            # Store docid in set of unindexed docids
+            self._not_indexed.add(docid)
+
+            return None
+
+        if isinstance(value, Persistent):
+            raise ValueError('Catalog cannot index persistent object %s' %
+                             value)
+
+        if isinstance(value, Broken):
+            raise ValueError('Catalog cannot index broken object %s' %
+                             value)
+
+        if docid in self._not_indexed:
+            # Remove from set of unindexed docs if it was in there.
+            self._not_indexed.remove(docid)
+
+        return self.inner_index_doc(docid, value)
+
+    def inner_index_doc(self, docid, value):
+        """See interface IInjection"""
+        rev_index = self._rev_index
+        if docid in rev_index:
+            docids = self._fwd_index.get(value, ())
+            if isinstance(docids, int):
+                if docids == docid:
+                    return
+            elif docid in docids:
+                # no need to index the doc, its already up to date
+                return
+            # unindex doc if present
+            self.unindex_doc(docid)
+
+        # Insert into forward index.
+        curdocids = self._fwd_index.get(value)
+        if curdocids is None:
+            self._fwd_index[value] = docid  # integer when only one docid
+        else:
+            if isinstance(curdocids, int):
+                curdocnum = 1
+            else:
+                curdocnum = len(curdocids)
+            newdocids = curdocids
+
+            if isinstance(curdocids, int):
+                newdocids = (curdocids,)
+            elif curdocnum == threshold-1 and isinstance(curdocids, tuple):
+                newset = self.family.IF.TreeSet()
+                newset.update(curdocids)
+
+                newdocids = newset
+
+            if isinstance(newdocids, tuple):
+                newdocids += (docid,)
+            elif isinstance(newdocids, self.family.IF.TreeSet):
+                newdocids.insert(docid)
+
+            self._fwd_index[value] = newdocids
```

Comment:
When newdocids is a TreeSet, you tell ZODB that you've changed _fwd_index's bucket. In fact, when newdocids was already a TreeSet, you didn't change the bucket.

So, you want to do `self._fwd_index[value] = newdocids` only if you had a tuple (`if isinstance(curdocids, tuple)`?)